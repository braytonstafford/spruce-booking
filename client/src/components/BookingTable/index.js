import React, { useState, useEffect } from 'react';
import { withStyles, makeStyles } from '@material-ui/core/styles';
import {
  Table,
  TableBody,
  TableCell,
  TableRow,
  TableHead,
  TableFooter,
  TablePagination,
} from '@material-ui/core';
import { grey } from '@material-ui/core/colors';
import moment from 'moment';

import BookingTableActions from '../BookingTableActions';

const useStyles = makeStyles((theme) => ({
  table: {
    marginLeft: 9,
  },
}));

const StyledTableCell = withStyles((theme) => ({
  head: {
    backgroundColor: grey[400],
    color: theme.palette.common.black,
  },
  body: {
    fontSize: 14,
  },
}))(TableCell);

const StyledTableRow = withStyles((theme) => ({
  root: {
    backgroundColor: theme.palette.common.white,
  },
}))(TableRow);

const BookingTable = () => {
  const classes = useStyles();
  const [rows, setRows] = useState([]);
  const [page, setPage] = React.useState(0);
  const [rowsPerPage, setRowsPerPage] = React.useState(20);

  useEffect(() => {
    const getBookings = async () => {
      try {
        const bookingsResponse = await fetch(
          'http://localhost:4000/api/booking',
          {
            method: 'GET',
            headers: { 'Content-Type': 'application/json' },
          }
        );

        const bookings = await bookingsResponse.json();
        setRows(
          bookings.sort(
            (a, b) => new Date(a.booking_date) - new Date(b.booking_date)
          )
        );
      } catch (err) {
        console.error('Failed to fetch bookings: ', err);
      }
    };

    getBookings();
  }, []);

  const handleChangePage = (event, newPage) => {
    setPage(newPage);
  };

  const handleChangeRowsPerPage = (event) => {
    setRowsPerPage(parseInt(event.target.value, 10));
    setPage(0);
  };

  return (
    <Table className={classes.table} aria-label='bookings table'>
      <TableHead>
        <TableRow>
          <StyledTableCell>Customer</StyledTableCell>
          <StyledTableCell align='right'>Email</StyledTableCell>
          <StyledTableCell align='right'>Address</StyledTableCell>
          <StyledTableCell align='right'>Booking Type</StyledTableCell>
          <StyledTableCell align='right'>Booking Date/Time</StyledTableCell>
        </TableRow>
      </TableHead>
      <TableBody>
        {(rowsPerPage > 0
          ? rows.slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)
          : rows
        ).map((row) => (
          <StyledTableRow key={row.booking_id}>
            <TableCell component='th' scope='row'>
              {row.first_name} {row.last_name}
            </TableCell>
            <TableCell align='right'>{row.email}</TableCell>
            <TableCell align='right'>{row.address1}</TableCell>
            <TableCell align='right'>{row.booking_type}</TableCell>
            <TableCell align='right'>
              {moment(row.booking_date).format('MMMM D, YYYY')} at{' '}
              {moment(row.booking_time).format('h:mm a')}
            </TableCell>
          </StyledTableRow>
        ))}
      </TableBody>
      <TableFooter>
        <TableRow>
          <TablePagination
            rowsPerPageOptions={[5, 10, 20, { label: 'All', value: -1 }]}
            colSpan={5}
            count={rows.length}
            rowsPerPage={rowsPerPage}
            page={page}
            SelectProps={{
              inputProps: { 'aria-label': 'rows per page' },
              native: true,
            }}
            onChangePage={handleChangePage}
            onChangeRowsPerPage={handleChangeRowsPerPage}
            ActionsComponent={BookingTableActions}
          />
        </TableRow>
      </TableFooter>
    </Table>
  );
};

export default BookingTable;
