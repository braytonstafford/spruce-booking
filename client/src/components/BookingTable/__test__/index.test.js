import React from 'react';
import BookingTable from '../index';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer.create(<BookingTable />).toJSON();
  expect(tree).toMatchSnapshot();
});
