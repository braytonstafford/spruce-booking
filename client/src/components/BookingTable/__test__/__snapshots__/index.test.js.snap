// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<table
  aria-label="bookings table"
  className="MuiTable-root makeStyles-table-1"
  role={null}
>
  <thead
    className="MuiTableHead-root"
    role={null}
  >
    <tr
      className="MuiTableRow-root MuiTableRow-head"
      role={null}
    >
      <th
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-2"
        scope="col"
      >
        Customer
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-2 MuiTableCell-alignRight"
        scope="col"
      >
        Email
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-2 MuiTableCell-alignRight"
        scope="col"
      >
        Address
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-2 MuiTableCell-alignRight"
        scope="col"
      >
        Booking Type
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-2 MuiTableCell-alignRight"
        scope="col"
      >
        Booking Date/Time
      </th>
    </tr>
  </thead>
  <tbody
    className="MuiTableBody-root"
    role={null}
  />
  <tfoot
    className="MuiTableFooter-root"
    role={null}
  >
    <tr
      className="MuiTableRow-root MuiTableRow-footer"
      role={null}
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-footer MuiTablePagination-root"
        colSpan={5}
      >
        <div
          className="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
        >
          <div
            className="MuiTablePagination-spacer"
          />
          <p
            className="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
          >
            Rows per page:
          </p>
          <div
            className="MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot"
            onClick={[Function]}
          >
            <select
              aria-label="rows per page"
              className="MuiSelect-root MuiSelect-select MuiTablePagination-select MuiInputBase-input"
              multiple={false}
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              value={20}
            >
              <option
                className="MuiTablePagination-menuItem"
                value={5}
              >
                5
              </option>
              <option
                className="MuiTablePagination-menuItem"
                value={10}
              >
                10
              </option>
              <option
                className="MuiTablePagination-menuItem"
                value={20}
              >
                20
              </option>
              <option
                className="MuiTablePagination-menuItem"
                value={-1}
              >
                All
              </option>
            </select>
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
          <p
            className="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
          >
            0-0 of 0
          </p>
          <div
            className="makeStyles-root-4"
          >
            <button
              aria-label="first page"
              className="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
              disabled={true}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M18.41 16.59L13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
                  />
                </svg>
              </span>
            </button>
            <button
              aria-label="previous page"
              className="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
              disabled={true}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
                  />
                </svg>
              </span>
            </button>
            <button
              aria-label="next page"
              className="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
              disabled={true}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
                  />
                </svg>
              </span>
            </button>
            <button
              aria-label="last page"
              className="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
              disabled={true}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M5.59 7.41L10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </td>
    </tr>
  </tfoot>
</table>
`;
